src/date.c

```diff
@@ -1,18 +1,1012 @@
+                                            &tp->min, &tp->sec, gmt, tz);
+    if(res == 3 || (res >= 6 && res <= 8) ) {
+      if( (tp->month = d_getnumbyname(mon_name,
+                                      ARRAY_SIZE(mon_name), month)) != -1) {
+        if(res == 7 && strncmp(gmt, "GMT", 3) == 0 ) {
+          *tz = 0;
+        }
+        return 1;
+      }
+    }
+  }
+  
+  /* #4: trying the rest */
+  
+  /* trying date */
+
+  if(!(date_parse(str, &tp->month, &tp->day, &tp->year, '/', buf1) >= 3 ||
+       date_parse(str, &tp->day, &tp->month, &tp->year, '.', buf1) >= 3 ||
+       date_parse(str, &tp->year, &tp->month, &tp->day, '-', buf1) >= 3 ) ) {
+    return 0;
+  }
+  
+  /*  there is date, trying time; from here return 0 only on errors */
+  
+  /* trying HH:mm */
+  {
+    const char* frmMMhh = " %d:%d%[^\0]";
+    res = sscanf(buf1, frmMMhh, &tp->hour, &tp->min, buf2);
+    /* can't get time, but have some symbols, assuming error */
+    if(res < 2 ) {
+      return (strlen(buf2) == 0);
+    }
+  }
+  
+  /* trying seconds */
+  {
+    const char* frmss = ":%d%[^\0]";
+    memset(buf1, 0, sizeof(buf1));
+    res = sscanf(buf2, frmss, &tp->sec, buf1);
+  }
+  
+  /* even if we don't get seconds we gonna try to get tz */
+  {
+    char* rest = res? buf1: buf2;
+    char* buf = res? buf2: buf1;
+    const char* frmtz = " %03s%d%[^\0]";
+    
+    res = sscanf(rest, frmtz, gmt, tz, buf);
+    if(res == 1 && strncmp(gmt, "GMT", 3) == 0) {
+      *tz = 0;
+    }
+  }
+
+  /* return OK if we are at the end of string */
+  return (res <= 2);
+}
+
+static int d_timeFromString(etime_t *time, const char *str, size_t buf_size) {
+  struct timeparts tp;
+  int tz;
+  
+  *time = INVALID_TIME;
+  
+  if(buf_size > 100) {
+    /* too long for valid date string */
+    return 0;
+  }
+  
+  if(d_parsedatestr(str, &tp, &tz)) {
+    /* check results */
+    int valid = 0;
+    
+    tp.month--;
+    valid = tp.day >=1 && tp.day <= 31;
+    valid &= tp.month >=0 && tp.month <= 11;
+    valid &= tp.hour >= 0 && tp.hour <= 23;
+    valid &= tp.min >= 0 && tp.min <= 59;
+    valid &= tp.sec >= 0 && tp.sec <= 59;
+    
+    if(tz != NO_TZ && tz > 12) {
+      tz /= 100;
+    }
+    
+    valid &= (abs(tz) <= 12 || tz == NO_TZ);
+    
+    if(valid) {
+      *time = d_gmktime(&tp);
+
+      if(tz != NO_TZ) {
+        /* timezone specified, use it */
+        *time -= tz * msPerHour;
+      } else {
+        /* assuming local timezone and moving back to UTC */
+        *time = ecma_UTC(*time);
+      }
+    }
+  }
+
+  return *time != INVALID_TIME;
+}
+
+typedef int (*ftostring)(struct timeparts*, char*, int);
+
+static val_t d_tostring(struct v7 *v7, val_t obj, fbreaktime breaktimefunc,
+                                        ftostring tostringfunc, int addtz) {
+  struct timeparts tp;
+  int len;
+  char buf[100];
+  etime_t time;
+  
+  d_checkobjvalidforstring(v7, obj);
+  
+  time = i_as_num(v7, obj);
```

Comment:
I got your idea, and walked through the code and fix functions with non-single i_value_of call. 
But I kept the very this place (d_tostring) as is, because it is called directly from Date API functions (ie macroses) and there isn't much sense to change it.